### ISSUE TITLE
Should the Host Language Label be used if it's empty?

### ISSUE THREAD

**Issue by dandclark** [w3c/accname#234]
The [Host Language Label step E](https://w3c.github.io/accname/#comp_host_language_label) of the accname computation algorithm is unclear on whether the empty string should be used if that's what the native markup provides.

> E. Host Language Label: Otherwise, if the current node's native markup provides an [attribute](https://dom.spec.whatwg.org/#concept-attribute) (e.g. alt) or [element](https://dom.spec.whatwg.org/#concept-element) (e.g. HTML label or SVG title) that defines a text alternative, return that alternative in the form of a flat string as defined by the host language, unless the element is marked as presentational (role="presentation" or role="none").

Going by a strict reading of the spec, I'd guess that an empty string should be used. However the `img with tooltip label with empty alt` subtest of [accname/name/comp_tooltip.html](https://wpt.fyi/results/accname/name/comp_tooltip.html?label=master&label=experimental&aligned&view=interop&q=label%3Ainterop-2024-accessibility) suggests that an empty `alt` attribute should be ignored. Gecko and WebKit pass this test, while Blink fails.

If that test reflects the intended behavior, should the spec be updated to something like this?

> E. Host Language Label: Otherwise, if the current node's native markup provides an [attribute](https://dom.spec.whatwg.org/#concept-attribute) (e.g. alt) or [element](https://dom.spec.whatwg.org/#concept-element) (e.g. HTML label or SVG title) that defines a text alternative, return that alternative in the form of a flat string as defined by the host language, unless the text alternative is an empty string or the element is marked as presentational (role="presentation" or role="none").

Also -- if the empty string is omitted, what about a whitespace-only string? (See https://github.com/w3c/accname/issues/233).

**Comment by scottaohara** [comment]
related https://github.com/w3c/html-aam/pull/533

**Comment by MelSumner** [comment]
The test reflects the desired behavior but `alt=""` is special TBH. 

I have long been of the opinion that, since modern IDEs will pre-fill `<img src="" alt=""/>` for you, it should be permissible/valid to also add `role="none"` or `role="presentation"` to an image. 

I do not think the specification should change here. If anything, we could add a note that this is a special case.

**Comment by MelSumner** [comment]
TL;DR: 
- Question: Should the host language label be used if it's empty?
- Answer: Yes. 

Here's the long answer.

If I'm understanding the question correctly, I think the question boils down to this: 

> if the native markup provides an empty string as a text alternative (e.g., an empty `alt=""` on `<img>`), should that empty string be used as the accessible name?


### Correct Interpretation: An Empty String Should Be Used

A strict reading of the spec suggests that "if the host language provides an empty string, it must be used as the accessible name".

---


### **Evidence Supporting This Interpretation**
1. **The Spec Says "Return That Alternative"**
   - Step E states:  
     > "return that alternative in the form of a flat string as defined by the host language."
   - If the host language (e.g., HTML) defines that an empty string (`""`) is a valid alternative text, then **returning `""` is the correct behavior**.

2. **HTML and ARIA Best Practices Align with This**
   - In HTML, `alt=""` **explicitly** conveys that an image should be ignored by screen readers.
   - The same logic applies to `<title>` elements inside SVGs:  
     ```html
     <svg>
       <title></title> <!-- Empty title -->
       <rect width="100" height="100" />
     </svg>
     ```
     Since the title exists but is empty, its contribution to the accessible name would be `""`, resulting in **no name for the SVG**.

3. **Precedent from ARIA `aria-labelledby=""`**
   - If `aria-labelledby` references an element that **exists but is empty**, the computed accessible name is an **empty string**, not `null`.  
   
     ```html
     <label id="label"></label>
     <input aria-labelledby="label">
     ```
     - Accessible name: `""` (empty string).
     - The element is not unnamed (`null`), but explicitly has **an empty string as its name**.

4. **Consistent Behavior with Host Language**
   - **HTML `<img alt="">` → Exposed as `name=""` (not `null`).**
   - **SVG `<title></title>` → Exposed as `name=""` (not `null`).**
   - **ARIA `aria-labelledby=""` → Exposed as `name=""` (not `null`).**

---

### **Why This Matters**
- **Empty strings signal intentional omission.** This is particularly important for images (`alt=""` means "decorative, ignore me").
- **Returning `null` vs. `""` changes behavior.** If the accessible name computation returned `null`, an element might fall back to other naming mechanisms (e.g., child text content), which would be incorrect.
- **Preserves fidelity with native behaviors.** The computed name should reflect what is actually present in the DOM, even if it is an empty string.

---

### **Conclusion**
✅ The **correct interpretation** of Step E is that **an empty string (`""`) must be used if the host language provides it**.  
🚫 The accessible name should **not be `null` or undefined** unless no alternative text mechanism is provided at all.

**Comment by rahimabdi** [comment]
I found out that the empty alt treatment is included in the HTML spec (https://html.spec.whatwg.org/multipage/embedded-content.html#the-img-element):

> What an [img](https://html.spec.whatwg.org/multipage/embedded-content.html#the-img-element) element represents depends on the [src](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-src) attribute and the [alt](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-alt) attribute.
> 
> **If the [src](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-src) attribute is set and the [alt](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-alt) attribute is set to the empty string**
> The image is either decorative or supplemental to the rest of the content, redundant with some other information in the document.

**Comment by rahimabdi** [comment]
@MelSumner I've created a PR for what appears to be an editorial change, but will defer to you if it resolves this issue 😄: https://github.com/w3c/aria/pull/2486.
